test/toolchain/toolchain_test.bzl

```diff
@@ -0,0 +1,128 @@
+"""End to end tests for rust toolchains."""
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest", "asserts")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("//rust:defs.bzl", "rust_library")
+load("//rust:toolchain.bzl", "rust_stdlib_filegroup", "rust_toolchain")
+
+EXEC_TOOLCHAIN_FLAG = "missing"
+TOOLCHAIN_FLAG = "before"
+CONFIG_FLAG = "after"
+
+def _transition_impl(settings, attr):
+    return {"//command_line_option:extra_toolchains": [":rust_extra_flags_toolchain"]}
+
+toolchain_transition = transition(
+    implementation = _transition_impl,
+    inputs = [],
+    outputs = ["//command_line_option:extra_toolchains"],
+)
+
+def _wrap_lib_impl(ctx):
+    pass
+
+wrap_lib = rule(
+    implementation = _wrap_lib_impl,
+    attrs = {
+        "dep": attr.label(cfg = toolchain_transition),
+        "_allowlist_function_transition": attr.label(default = "@bazel_tools//tools/allowlists/function_transition_allowlist"),
+    },
+)
+
+def _toolchain_adds_rustc_flags_impl(ctx):
+    env = analysistest.begin(ctx)
+
+    target = ctx.attr.lib
+    action = target.actions[0]
+    asserts.equals(env, "Rustc", action.mnemonic)
+
+    asserts.true(
+        env,
+        action.argv[-2:] == [TOOLCHAIN_FLAG, CONFIG_FLAG],
+        "Unexpected rustc flags: {}\nShould have ended with: {}".format(
+            action.argv,
+            [TOOLCHAIN_FLAG, CONFIG_FLAG],
+        ),
+    )
+
+    asserts.true(
+        env,
+        EXEC_TOOLCHAIN_FLAG not in action.argv,
+        "Found exec toolchain flag ({}) in rustc flags: {}".format(EXEC_TOOLCHAIN_FLAG, action.argv),
+    )
+
+    return analysistest.end(env)
+
+toolchain_adds_rustc_flags_test = analysistest.make(
+    _toolchain_adds_rustc_flags_impl,
+    attrs = {
+        "lib": attr.label(mandatory = True),
+    },
+    config_settings = {
+        "@//:extra_rustc_flags": [CONFIG_FLAG],
+        # "//command_line_option:extra_toolchains": [":rust_extra_flags_toolchain"],
```

Comment:
I was also curious if we could avoid the transition by specifying the extra toolchain directly here, but I run into the same issue where it errors because the toolchain hasn't been declared yet.

Reply:
as it turns out: yes we can use the `config_settings` from `analysistest` to avoid having to make the transition/wrapper rule.